### PR TITLE
Autodiff gradient formula for addmm

### PR DESCRIPTION
**Pull request by jamesr66a**

### CODE REVIEW

File: torch/csrc/jit/autodiff.cpp [hunk 1]
```diff
@@ -149,6 +159,32 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
           return returned_grads;
         }
       }
+      case aten::addmm: {
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/autodiff.cpp]: The fission pass has to go in specializeToSpec, to run _before_ shape propagation, so that expands get correctly inserted.

**Comment by jekbradbury** [comment]
Is there an eventual goal for these `SymbolicVariable` expressions to also be generated from derivatives.yaml? Or does that not have (even in principle) the information that JIT AD needs?

File: torch/csrc/jit/autodiff.cpp [hunk 2]
```diff
@@ -149,6 +159,32 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
...
+        JIT_ASSERT(inputs.size() == 3);
+        JIT_ASSERT(outputs.size() == 1);
+        JIT_ASSERT(grads.size() == 1);
```
Comment (apaszke) [inline torch/csrc/jit/autodiff.cpp]: I think you can assume that `outputs().size() == grads.size()` at this point. We dont want to check this in every single derivative formula

File: torch/csrc/jit/autodiff.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Wouldn't it be simpler to just do that instead of this PR? It's generally a useful and simple pass, and it would simplify the change to AD

File: torch/csrc/jit/autodiff.cpp [hunk 3]
```diff
@@ -17,8 +17,12 @@ bool isDifferentiable(Node * n) {
   static std::unordered_set<Symbol> differentiable_kinds = {
     aten::add, aten::sub, aten::mul, prim::Constant, prim::ReplaceIfUndef,
     aten::sigmoid, aten::tanh, aten::mm, aten::chunk, aten::split, aten::t, aten::neg,
-    aten::unsqueeze, aten::expand,
+    aten::unsqueeze, aten::expand, aten::addmm
   };
+  // alpha and beta parameters must be attributes
+  if (n->kind() == aten::addmm && n->inputs().size() > 3) {
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Wouldn't this conflict with out canonicalization efforts?

File: test/test_jit.py [hunk 4]
```diff
@@ -2714,6 +2714,30 @@ def t2(a):
...
 
+    def test_addmm_grad(self):
+        class LinearTest(torch.jit.ScriptModule):
+            def __init__(self, param_init):
+                super(LinearTest, self).__init__()
+                self.param = torch.nn.Parameter(param_init)
+
+            @torch.jit.script_method
+            def forward(self, x):
+                return torch.addmm(x, x, self.param)
```
Comment (apaszke) [inline test/test_jit.py]: Wouldn't it be simpler to just have a function that does `addmm` inside? There's nothing specific to modules in this test, and it would greatly simplify it.

File: test/test_jit.py [hunk 5]
```diff
@@ -2714,6 +2714,30 @@ def t2(a):
...
+
+        param_init = torch.rand(2, 2)
+        lt = LinearTest(param_init)
+        x = torch.rand(2, 2)
+        y = lt(x)
+        y.sum().backward()
+        grad_test = lt.param.grad
```
Comment (apaszke) [inline test/test_jit.py]: A nicer fix really would be to make `checkScript` check the correctness of gradients (analogously to `checkTrace`).

File: test/test_jit.py [hunk 6]
```diff
@@ -2714,6 +2714,30 @@ def t2(a):
...
+
+        param_test = torch.autograd.Variable(param_init, requires_grad=True)
```
Comment (apaszke) [inline test/test_jit.py]: Pls no Variables... Use `param_init.requires_grad_()`

File: test/test_jit.py [hunk 7]
```diff
@@ -2714,6 +2714,30 @@ def t2(a):
...
+        y_ref = torch.addmm(x, x, param_test)
+        y_ref.sum().backward()
+        grad_ref = param_test.grad
```
Comment (apaszke) [inline test/test_jit.py]: This line is really unnecessary. You use it once in an assert

File: torch/csrc/jit/autodiff.cpp [hunk 8]
```diff
@@ -149,6 +159,32 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
...
+          }
+        }
+
+        // Grad from the addition
+        auto beta = at::Scalar(node->t(attr::beta));
+        auto it = returned_grads.begin();
+        if (beta.to<double>() != 1.0)
+          returned_grads.insert(it, grads[0] * beta);
+        else
+          returned_grads.insert(it, grads[0]);
```
Comment (apaszke) [inline torch/csrc/jit/autodiff.cpp]: Soo this is expected to work only under the assumption that the add part doesn't broadcast, right?

**Comment by apaszke** [comment]
@jekbradbury I don't think so. The dependencies go the other way around - `SymbolicVariable` is something that just makes building graphs easier (in particular in `autodiff.cpp`), not something that actually knows how to differentiate itself. It should probably be called `SymbolicTensor` 😄

**Comment by jekbradbury** [comment]
Yes, I guess what I mean is that the expressions in derivatives.yaml are basically symbolic; I'm wondering what prevents using them to generate the gradient-graph-building code currently handwritten in autodiff.c

File: torch/csrc/jit/autodiff.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/csrc/jit/autodiff.cpp]: One common example of a*b+c is calculating grads of linear layer params, followed by AccumulateGrad - is AccumulateGrad in the scope of the jit tracer?

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Well it's never traced at the moment, so no. In the future it might be, and we'll need to take care of that.

File: torch/csrc/jit/autodiff.cpp [hunk 9]
```diff
@@ -28,6 +32,33 @@ bool isDifferentiable(Graph & g) {
...
+  }
+  if (auto type = inputs.at(1).value()->type()->cast<TensorType>()) {
+    auto sizes = type->sizes(), strides = type->strides();
+    if (strides.at(0) == 1 && strides.at(1) == sizes.at(0)) {
+      dmat2 = grads.at(0).t().mm(inputs.at(0)).t();
+    } else {
+      dmat2 = inputs.at(0).t().mm(grads.at(0));
+    }
+  } else {
+    dmat2 = grads.at(0).mm(inputs.at(1).t());
```
Comment (ngimel) [inline torch/csrc/jit/autodiff.cpp]: There's a typo here (should be `dmat2 = inputs.at(0).t().mm(grads.at(0));`), the same typo is in the current version in autodiff.cpp.

**Comment by jamesr66a** [comment]
Superseding this with https://github.com/pytorch/pytorch/issues/7930. PRs to come
